keps/sig-cli/2953-kustomize-plugin-graduation/README.md

```diff
@@ -0,0 +1,764 @@
+
+### Plugin configuration
+
+All plugins must be configured using KRM-style custom-resource-like objects.
+
+The `config.kubernetes.io/function` annotation that currently contains a nested JSON blob of plugin provider configuration will be graduated to a reserved top-level field called `provider` (ALTERNATIVE: it could be nested as `metadata.provider`, but that would make the object meta non-standard). That field will be made optional by the [Catalog KEP].
+
+The `metadata.name` field is optional as long as the config's GVK uniquely identifies it. In Kustomization, it is not required unless the config is being handled as a resource. In Composition, it is always recommended, because transformer config is always handled as a resource before execution and uniqueness must be global across all Compositions in the tree. 
+
+The `env` field for container plugins will be deprecated. In line with Kustomize's principles, plugin configuration should be passed in the KRM object and should not be affected by the execution environment. The network and storage mount options will be retained, as they are needed to support some common generator plugin use cases. Guidance on their appropriate use will be added to the plugin developer documentation.
+
+### Legacy plugin migration
+
+Although legacy plugins are still in alpha, the alpha has been around in standalone Kustomize for quite a while. Documentation on the migration process will be provided. See the [Rollout Plan](#rollout-plan) section for more detail on deprecation and removal of the existing alphas.
+
+The exec plugin conversion process should be relatively straightforward: copy the provider in-tree and update the configuration that referred to it to have a Provider stanza pointing to the new location. `argsFromFile` and `argsOneLiner` can be straightforwardly converted to the `args` subfield the KRM Functions style already supports.
+
+For authors who build plugins for distribution (as opposed to bespoke plugins where embedding within an end-user Kustomization is an appropriate solution), the recommendation will be to upgrade to the Catalog model. A Catalog guide will be added to the Kustomize website and will address the scenario of migrating a legacy style plugin.
+
+
+
+### User Stories (Optional)
+
+#### Story 1
+
+As an end user, I want to develop a simple plugin for my particular Kustomization.
+
+1. Write a script that adheres to the [KRM Functions Specification] and place it within the Kustomization root.
+2. Write a corresponding plugin config and reference it from the Kustomization's generator or transformer field as appropriate.
+3. Run `kustomize edit generate-catalog`
```

Comment:
Is this necessary? Would it be possible to just run `kustomize build` without generating the catalog?

Reply:
The reason for not allowing that is that this proposal uses Catalog as the sole trust model. For any solution, we need the end user (or process running the command) to be telling us that they're aware Kustomize is going to invoke third-party code, and ideally _which_ third party code they are ok with invoking. We could introduce a separate mechanism for ad-hoc local plugins, but after writing out the two alternatives presented here, I felt it was probably simpler for users if there is one consistent workflow that we invest in making easy for folks to follow (i.e. by providing the command in this step).

Reply:
also a fan of fewer commands.  Onboarding kpt users and writing end user docs showed me that Solution Architects will ask for every possible command to be reduced because (a) it introduces friction (b) it introduces opportunities for things to get out of sync.  At the end of the day someone needs to prove to a platform admin why doing things through kustomize is superior to just runnings sorter.rb directly granted that their CICD image can add a locked down version or Ruby.